In [20]:
import pandas as pd
import numpy as np
# Reset the 'display.max_colwidth' option to its default value
pd.reset_option('display.max_colwidth')


In [21]:
df = pd.read_csv("QA_Training.realestate.csv")
df.head()

_id               first_name middle_name   last_name  \
0  66cbff36e46b6066699d13b5  Gregg & Debbie Boosalis         NaN         NaN   
1  66cbff36e46b6066699d13b6                      Mai         NaN          Do   
2  66cbff36e46b6066699d13b7                     Jeff         NaN      Searcy   
3  66cbff36e46b6066699d13b8                      Lea         Ann  Stephenson   
4  66cbff36e46b6066699d13b9                  Jeanine         NaN      Curtis   

             title                                          image_url  \
0     Team Website  https://content.mediastg.net/Dynamic/RealEstat...   
1          Trainer  https://content.mediastg.net/Dynamic/RealEstat...   
2  The Searcy Team  https://content.mediastg.net/Dynamic/RealEstat...   
3         REALTOR®  https://content.mediastg.net/Dynamic/RealEstat...   
4  Sales Associate  https://content.mediastg.net/Dynamic/RealEstat...   

  agent_phone_numbers[0] office_phone_numbers[0]                  office_name  \
0           402-432-3666            402-436-4663  HOME Real Estate At Midtown   
1             4023049394            402-436-4663  HOME Real Estate At Midtown   
2           402-430-7978                     NaN  HOME Real Estate At Midtown   
3           402-506-3875                     NaN  HOME Real Estate At Midtown   
4           402-429-1200                     NaN  HOME Real Estate At Midtown   

         address  ... zipcode  \
0  5322 O Street  ...   68510   
1  5322 O Street  ...   68510   
2  5322 O Street  ...   68510   
3  5322 O Street  ...   68510   
4  5322 O Street  ...   68510   

                                         description languages[0]  \
0  As life long residents of Lincoln, we have bee...      Russian   
1  As part of HOME Real Estate, we have a whole t...          NaN   
2  Greetings!Approaching 50 years of service to o...          NaN   
3  My WHYto help people and leave the world bette...          NaN   
4  As part of HOME Real Estate, we have a whole t...          NaN   

         country                                social.facebook_url  \
0  United States                                                NaN   
1  United States                                                NaN   
2  United States  https://www.facebook.com/SearcyTeamHomeRealEstate   
3  United States  https://www.facebook.com/LeaAnnStephensonRealE...   
4  United States                                                NaN   

  social.twitter_url                                social.linkedin_url  \
0                NaN                                                NaN   
1                NaN                                                NaN   
2                NaN                                                NaN   
3                NaN  https://www.linkedin.com/in/lea-ann-stephenson...   
4                NaN                                                NaN   

                                         profile_url  \
0  https://gregganddebbieboosalis.homerealestate.com   
1                   https://maido.homerealestate.com   
2              https://JeffSearcy.homerealestate.com   
3        https://leaannstephenson.homerealestate.com   
4           https://JeanineCurtis.homerealestate.com   

              social.other_urls[0]                      social.other_urls[1]  
0                              NaN                                       NaN  
1                              NaN                                       NaN  
2                              NaN                                       NaN  
3  http://blog.homerealestate.com/  https://www.instagram.com/annie_steph07/  
4                              NaN                                       NaN  

[5 rows x 24 columns]

In [22]:
df.columns

Index(['_id', 'first_name', 'middle_name', 'last_name', 'title', 'image_url',
       'agent_phone_numbers[0]', 'office_phone_numbers[0]', 'office_name',
       'address', 'email', 'website', 'city', 'state', 'zipcode',
       'description', 'languages[0]', 'country', 'social.facebook_url',
       'social.twitter_url', 'social.linkedin_url', 'profile_url',
       'social.other_urls[0]', 'social.other_urls[1]'],
      dtype='object')

In [23]:
df.shape

(161, 24)

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 161 entries, 0 to 160
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   _id                      161 non-null    object
 1   first_name               161 non-null    object
 2   middle_name              8 non-null      object
 3   last_name                153 non-null    object
 4   title                    160 non-null    object
 5   image_url                152 non-null    object
 6   agent_phone_numbers[0]   157 non-null    object
 7   office_phone_numbers[0]  56 non-null     object
 8   office_name              161 non-null    object
 9   address                  161 non-null    object
 10  email                    160 non-null    object
 11  website                  161 non-null    object
 12  city                     161 non-null    object
 13  state                    161 non-null    object
 14  zipcode                  161 non-null    o

In [25]:
df.describe().transpose()

count unique  \
_id                       161    161   
first_name                161    141   
middle_name                 8      7   
last_name                 153    135   
title                     160     20   
image_url                 152    152   
agent_phone_numbers[0]    157    149   
office_phone_numbers[0]    56      3   
office_name               161      2   
address                   161      2   
email                     160    157   
website                   161    161   
city                      161      2   
state                     161      1   
zipcode                   161      3   
description               145     64   
languages[0]               59      5   
country                   161      1   
social.facebook_url        59     50   
social.twitter_url          2      1   
social.linkedin_url        15     14   
profile_url               161    161   
social.other_urls[0]       11      9   
social.other_urls[1]        3      3   

                                                                       top  \
_id                                               66cbff36e46b6066699d13b5   
first_name                                                             Jim   
middle_name                                                              &   
last_name                                                           Searcy   
title                                                      Sales Associate   
image_url                https://content.mediastg.net/Dynamic/RealEstat...   
agent_phone_numbers[0]                                        402-643-3300   
office_phone_numbers[0]                                       402-436-4663   
office_name                                    HOME Real Estate At Midtown   
address                                                      5322 O Street   
email                                       andie.young@homerealestate.com   
website                  https://gregganddebbieboosalis.homerealestate.com   
city                                                              Lincoln    
state                                                                   NE   
zipcode                                                              68510   
description              As part of HOME Real Estate, we have a whole t...   
languages[0]                                                       Russian   
country                                                      United States   
social.facebook_url      https://www.facebook.com/abbyburmeisterandtiff...   
social.twitter_url                          https://twitter.com/@Abby_Tiff   
social.linkedin_url               https://www.linkedin.com/in/katieulrich/   
profile_url              https://gregganddebbieboosalis.homerealestate.com   
social.other_urls[0]     https://www.instagram.com/andie.jessi_realestate/   
social.other_urls[1]              https://www.instagram.com/annie_steph07/   

                        freq  
_id                        1  
first_name                 3  
middle_name                2  
last_name                  4  
title                     99  
image_url                  1  
agent_phone_numbers[0]     3  
office_phone_numbers[0]   51  
office_name              159  
address                  159  
email                      3  
website                    1  
city                     159  
state                    161  
zipcode                  158  
description               59  
languages[0]              20  
country                  161  
social.facebook_url        4  
social.twitter_url         2  
social.linkedin_url        2  
profile_url                1  
social.other_urls[0]       2  
social.other_urls[1]       1

In [26]:
df.isnull().sum()

_id                          0
first_name                   0
middle_name                153
last_name                    8
title                        1
image_url                    9
agent_phone_numbers[0]       4
office_phone_numbers[0]    105
office_name                  0
address                      0
email                        1
website                      0
city                         0
state                        0
zipcode                      0
description                 16
languages[0]               102
country                      0
social.facebook_url        102
social.twitter_url         159
social.linkedin_url        146
profile_url                  0
social.other_urls[0]       150
social.other_urls[1]       158
dtype: int64

In [27]:
# Show rows with missing values
df[df.isnull().any(axis=1)]

_id               first_name middle_name  \
0    66cbff36e46b6066699d13b5  Gregg & Debbie Boosalis         NaN   
1    66cbff36e46b6066699d13b6                      Mai         NaN   
2    66cbff36e46b6066699d13b7                     Jeff         NaN   
3    66cbff36e46b6066699d13b8                      Lea         Ann   
4    66cbff36e46b6066699d13b9                  Jeanine         NaN   
..                        ...                      ...         ...   
156  66cbff36e46b6066699d1451                   Rachel     Roberts   
157  66cbff36e46b6066699d1452                   Marsha         NaN   
158  66cbff36e46b6066699d1453                    Jason         NaN   
159  66cbff36e46b6066699d1454                     Jeff         NaN   
160  66cbff36e46b6066699d1455                   Alicia         NaN   

      last_name            title  \
0           NaN     Team Website   
1            Do          Trainer   
2        Searcy  The Searcy Team   
3    Stephenson         REALTOR®   
4        Curtis  Sales Associate   
..          ...              ...   
156      Spring         REALTOR®   
157        Artz  Sales Associate   
158      Nelson         REALTOR®   
159     Johnson  Sales Associate   
160     Prewett        REALTOR ®   

                                             image_url agent_phone_numbers[0]  \
0    https://content.mediastg.net/Dynamic/RealEstat...           402-432-3666   
1    https://content.mediastg.net/Dynamic/RealEstat...             4023049394   
2    https://content.mediastg.net/Dynamic/RealEstat...           402-430-7978   
3    https://content.mediastg.net/Dynamic/RealEstat...           402-506-3875   
4    https://content.mediastg.net/Dynamic/RealEstat...           402-429-1200   
..                                                 ...                    ...   
156  https://content.mediastg.net/Dynamic/RealEstat...           402-440-8129   
157  https://content.mediastg.net/Dynamic/RealEstat...           402-525-1215   
158  https://content.mediastg.net/Dynamic/RealEstat...           402-219-2505   
159  https://content.mediastg.net/Dynamic/RealEstat...           402-432-0756   
160  https://content.mediastg.net/Dynamic/RealEstat...           402-432-5864   

    office_phone_numbers[0]                  office_name        address  ...  \
0              402-436-4663  HOME Real Estate At Midtown  5322 O Street  ...   
1              402-436-4663  HOME Real Estate At Midtown  5322 O Street  ...   
2                       NaN  HOME Real Estate At Midtown  5322 O Street  ...   
3                       NaN  HOME Real Estate At Midtown  5322 O Street  ...   
4                       NaN  HOME Real Estate At Midtown  5322 O Street  ...   
..                      ...                          ...            ...  ...   
156                     NaN  HOME Real Estate At Midtown  5322 O Street  ...   
157                     NaN  HOME Real Estate At Midtown  5322 O Street  ...   
158                     NaN  HOME Real Estate At Midtown  5322 O Street  ...   
159                     NaN  HOME Real Estate At Midtown  5322 O Street  ...   
160                     NaN  HOME Real Estate At Midtown  5322 O Street  ...   

    zipcode                                        description languages[0]  \
0     68510  As life long residents of Lincoln, we have bee...      Russian   
1     68510  As part of HOME Real Estate, we have a whole t...          NaN   
2     68510  Greetings!Approaching 50 years of service to o...          NaN   
3     68510  My WHYto help people and leave the world bette...          NaN   
4     68510  As part of HOME Real Estate, we have a whole t...          NaN   
..      ...                                                ...          ...   
156   68510  As part of HOME Real Estate, we have a whole t...          NaN   
157   68510  Real Estate is a complex world. While there wi...          NaN   
158   68510  Hello!As part of HOME Real Estate, we have a w...          NaN   
159   68510  Jeff's immense 

In [29]:
df.duplicated().sum()

np.int64(0)

## First name

In [28]:
# Step 1: Find duplicate first names
duplicate_names = df[df.duplicated(subset='first_name', keep=False)]

# Step 2: Print rows with duplicate first names
print(duplicate_names)

                          _id               first_name middle_name  last_name  \
2    66cbff36e46b6066699d13b7                     Jeff         NaN     Searcy   
5    66cbff36e46b6066699d13ba                    Daria           &     Marsha   
14   66cbff36e46b6066699d13c3                     Nick         NaN      Hardt   
24   66cbff36e46b6066699d13cd                   Debbie         NaN   Messineo   
33   66cbff36e46b6066699d13d6                   Ashley         NaN  Rockemann   
37   66cbff36e46b6066699d13da                      Pat         NaN  Ohmberger   
41   66cbff36e46b6066699d13de                    Karen         NaN  Schaeffer   
43   66cbff36e46b6066699d13e0                    Katie         NaN     Bugbee   
49   66cbff36e46b6066699d13e6                    Daria         NaN   Springer   
58   66cbff36e46b6066699d13ef                   Connie         NaN    Boender   
63   66cbff36e46b6066699d13f4  Randy and Roma Amundson         NaN        NaN   
64   66cbff36e46b6066699d13f

In [30]:
# Identify rows where the 'first_name' column contains multiple words (indicating full names)
full_name_rows = df[df['first_name'].str.split().str.len() > 1]

# Display these rows
print("Rows with full names in the 'first_name' column:")
print(full_name_rows[['first_name']])

Rows with full names in the 'first_name' column:
                              first_name
0                Gregg & Debbie Boosalis
15   HOME Real Estate Customer Concierge
23        Kim Marie Schrader Eberspacher
61                   Katie & Brad Ulrich
63               Randy and Roma Amundson
90               Roma and Randy Amundson
108      Abby Burmeister & Tiffany Heier
141              Randy and Roma Amundson


In [32]:
df[(df['first_name']=='Randy and Roma Amundson') | (df['first_name'] == 'Roma and Randy Amundson')]

_id               first_name middle_name last_name  \
63   66cbff36e46b6066699d13f4  Randy and Roma Amundson         NaN       NaN   
90   66cbff36e46b6066699d140f  Roma and Randy Amundson         NaN       NaN   
141  66cbff36e46b6066699d1442  Randy and Roma Amundson         NaN       NaN   

                       title  \
63   Randy and Roma Amundson   
90           Sales Associate   
141          Sales Associate   

                                             image_url agent_phone_numbers[0]  \
63   https://content.mediastg.net/Dynamic/RealEstat...           402-432-7644   
90   https://content.mediastg.net/Dynamic/RealEstat...           402-432-5488   
141  https://content.mediastg.net/Dynamic/RealEstat...           402-432-7644   

    office_phone_numbers[0]                  office_name        address  ...  \
63             402-436-4663  HOME Real Estate At Midtown  5322 O Street  ...   
90                      NaN  HOME Real Estate At Midtown  5322 O Street  ...   
141            402-436-4663  HOME Real Estate At Midtown  5322 O Street  ...   

    zipcode                                        description languages[0]  \
63    68510  Our real estate service extends beyond merely ...          NaN   
90    68510  Our real estate service extends beyond merely ...          NaN   
141   68510  Our real estate service extends beyond merely ...          NaN   

           country                   social.facebook_url social.twitter_url  \
63   United States  https://www.facebook.com/RealtorRoma                NaN   
90   United States  https://www.facebook.com/RealtorRoma                NaN   
141  United States                                   NaN                NaN   

                                   social.linkedin_url  \
63                                                 NaN   
90   https://www.linkedin.com/in/roma-amundson-246b...   
141                                                NaN   

                                  profile_url social.other_urls[0]  \
63    https://RandyandRoma.homerealestate.com                  NaN   
90    https://romaamundson.homerealestate.com                  NaN   
141  https://randyamundson.homerealestate.com                  NaN   

    social.other_urls[1]  
63                   NaN  
90                   NaN  
141                  NaN  

[3 rows x 24 columns]

The above three are different websites but contain same info with slight differences. like one has social media details and the other doesnt have it.

In [34]:
# Step 1: Find duplicate middle names, excluding rows with null values
duplicate_names = df[df.duplicated(subset='middle_name', keep=False) & ~df['middle_name'].isna()]

# Step 2: Print rows with duplicate middle names
print(duplicate_names)

                          _id first_name middle_name last_name         title  \
5    66cbff36e46b6066699d13ba      Daria           &    Marsha  Team Website   
155  66cbff36e46b6066699d1450      Andie           &     Jessi  Team Website   

                                             image_url agent_phone_numbers[0]  \
5    https://content.mediastg.net/Dynamic/RealEstat...           402-525-1215   
155  https://content.mediastg.net/Dynamic/RealEstat...           402-326-4921   

    office_phone_numbers[0]                  office_name        address  ...  \
5              402-436-4663  HOME Real Estate At Midtown  5322 O Street  ...   
155            402-436-4663  HOME Real Estate At Midtown  5322 O Street  ...   

    zipcode                                        description languages[0]  \
5     68510                                                NaN          NaN   
155   68510  We could not be more thrilled to bring you top...          NaN   

           country                  

## Last Name

In [35]:
# Step 1: Find duplicate last names, excluding rows with null values
duplicate_names = df[df.duplicated(subset='last_name', keep=False) & ~df['last_name'].isna()]

# Step 2: Print rows with duplicate last names
print(duplicate_names)

                          _id first_name middle_name  last_name  \
2    66cbff36e46b6066699d13b7       Jeff         NaN     Searcy   
7    66cbff36e46b6066699d13bc     Jasper         NaN     Searcy   
10   66cbff36e46b6066699d13bf       Zach         NaN    Friesen   
17   66cbff36e46b6066699d13c6    Chelsea         NaN      Davis   
19   66cbff36e46b6066699d13c8     Dmitry         NaN      Rybak   
25   66cbff36e46b6066699d13ce     Alyssa         NaN     Osborn   
26   66cbff36e46b6066699d13cf     Hannah         NaN     Searcy   
30   66cbff36e46b6066699d13d3     Marion         NaN    Polivka   
31   66cbff36e46b6066699d13d4       Lydi         NaN     Gorbun   
32   66cbff36e46b6066699d13d5        Val         NaN   Javorsky   
35   66cbff36e46b6066699d13d8   Shawndel         NaN    Polivka   
37   66cbff36e46b6066699d13da        Pat         NaN  Ohmberger   
48   66cbff36e46b6066699d13e5        The      Searcy       Team   
51   66cbff36e46b6066699d13e8       Todd         NaN     Loren

## Title

In [36]:
df['title'].value_counts()

title
Sales Associate                           99
REALTOR®                                  31
Team Website                               7
The Searcy Team                            4
Associate Broker                           3
Manager                                    2
Trainer                                    1
Sales Agent                                1
Listing Coordinator                        1
REALTOR® | Associate Broker | CRS, GRI     1
Randy and Roma Amundson                    1
Customer Concierge                         1
ABR, GRI, PSA, SRS                         1
Office-Group Administrator                 1
REALTOR                                    1
REALTOR (R)                                1
REALTOR®, ABR, AHWD, C2EX, e-PRO           1
Abby Burmeister & Tiffany Heier            1
Relocation Director                        1
REALTOR ®                                  1
Name: count, dtype: int64

1. **Inconsistent Special Characters:**
   - Different variations of "REALTOR" with inconsistent use of special characters (e.g., ® and ® |).

2. **Inconsistent Formatting:**
   - Similar roles described differently (e.g., "Sales Associate" vs. "Sales Agent").
   - Multiple roles combined in one title (e.g., "REALTOR® | Associate Broker | CRS, GRI").

3. **Redundant Titles:**
   - Duplicate titles with slight variations (e.g., "REALTOR" vs. "REALTOR®").

4. **Potential Typos or Mislabeling:**
   - Names listed as titles (e.g., "Abby Burmeister & Tiffany Heier").

5. **Special Cases:**
   - Less common titles that might have varying definitions (e.g., "Office-Group Administrator").

## Image URL

In [73]:
import pandas as pd
import re

def url_qa(df, column_name):
    """
    Perform QA checks on a specified URL column in the DataFrame.

    Parameters:
    - df: DataFrame containing the URL column.
    - column_name: Name of the column to check.

    Returns:
    - A dictionary with DataFrames containing results of the QA checks.
    """
    # Ensure the column exists
    if column_name not in df.columns:
        raise ValueError(f"Column '{column_name}' does not exist in the DataFrame.")
    
    # Convert non-string values to empty strings
    df[column_name] = df[column_name].astype(str)
    
    # Regular expression for validating URLs
    url_regex = re.compile(r'^(https?|ftp)://[^\s/$.?#].[^\s]*$', re.IGNORECASE)
    
    # Function to validate URL format
    def is_valid_url(url):
        return bool(url_regex.match(url))
    
    # Add validation columns
    df[f'{column_name}_valid'] = df[column_name].apply(lambda x: is_valid_url(x) if pd.notna(x) else False)
    df[f'{column_name}_missing_protocol'] = ~df[column_name].str.contains('^https?://', na=False)
    df[f'{column_name}_invalid_chars'] = df[column_name].str.contains(r'[^a-zA-Z0-9:/._\-~]', regex=True, na=False)
    df[f'{column_name}_long'] = df[column_name].str.len() > 2048  # Example length limit
    
    # Identify issues
    missing_protocols = df[df[f'{column_name}_missing_protocol']][[column_name]]
    invalid_chars = df[df[f'{column_name}_invalid_chars']][[column_name]]
    long_urls = df[df[f'{column_name}_long']][[column_name]]
    invalid_urls = df[~df[f'{column_name}_valid']][[column_name]]
    
    # Collect results
    results = {
        'Missing Protocols': missing_protocols,
        'Invalid Characters': invalid_chars,
        'Long URLs': long_urls,
        'Invalid URLs': invalid_urls
    }
    
    return results


In [74]:
image_url_qa_results = url_qa(df, 'image_url')

print("\nImage URL QA Results:")
for key, result_df in image_url_qa_results.items():
    print(f"\n{key}:")
    # Exclude rows where 'image_url' is NaN
    print(result_df.dropna(subset=['image_url']))


Image URL QA Results:

Missing Protocols:
    image_url
11        nan
25        nan
30        nan
44        nan
50        nan
70        nan
118       nan
129       nan
138       nan

Invalid Characters:
Empty DataFrame
Columns: [image_url]
Index: []

Long URLs:
Empty DataFrame
Columns: [image_url]
Index: []

Invalid URLs:
    image_url
11        nan
25        nan
30        nan
44        nan
50        nan
70        nan
118       nan
129       nan
138       nan


## Agent Phone Numbers

In [75]:
df['agent_phone_numbers[0]'].value_counts()

agent_phone_numbers[0]
402-643-3300    3
402-525-1215    2
402-432-7644    2
402-480-1727    2
402-432-3666    2
               ..
402-429-5342    1
402-440-8129    1
402-219-2505    1
402-432-0756    1
402-432-5864    1
Name: count, Length: 149, dtype: int64

Is in object data type

## Office Phone Numbers

In [76]:
df['office_phone_numbers[0]'].value_counts()

office_phone_numbers[0]
402-436-4663      51
(402) 436-4663     3
402-434-3700       2
Name: count, dtype: int64

Inconsistent Formatting:
402-436-4663 and (402) 436-4663: These represent the same phone number in different formats. The parentheses and hyphens are used inconsistently.

Object Data Type:
Since the column is of type object, it means the phone numbers are stored as strings. This type is flexible but requires standardization for consistency.

Null values:
Has plenty of null values

## Office Name

In [77]:
df['office_name'].value_counts()

office_name
HOME Real Estate At Midtown          159
HOME Real Estate Support Services      2
Name: count, dtype: int64

In [78]:
# Group by 'office_name' and 'office_phone_numbers[0]' and count occurrences
df.groupby(['office_name', 'office_phone_numbers[0]']).size().reset_index(name='count')

office_name office_phone_numbers[0]  count
0        HOME Real Estate At Midtown          (402) 436-4663      3
1        HOME Real Estate At Midtown            402-436-4663     51
2  HOME Real Estate Support Services            402-434-3700      2

We can use this to fill null values in office_numbers column

## Address

In [79]:
df['address'].value_counts()

address
5322 O Street              159
7811 Pioneers Blvd #200      2
Name: count, dtype: int64

## Email

In [80]:
df['email'].value_counts()

email
andie.young@homerealestate.com          3
UnderwoodTeam@HomeRealEstate.com        2
jeff@searcyteam.com                     1
leaann.stephenson@homerealestate.com    1
jeanine.curtis@homerealestate.com       1
                                       ..
rachel.spring@homerealestate.com        1
marsha.artz@homerealestate.com          1
jason.nelson@homerealestate.com         1
jeff.johnson@homerealestate.com         1
alicia.prewett@homerealestate.com       1
Name: count, Length: 158, dtype: int64

In [81]:
df[df['email'].isna()]

Empty DataFrame
Columns: [_id, first_name, middle_name, last_name, title, image_url, agent_phone_numbers[0], office_phone_numbers[0], office_name, address, email, website, city, state, zipcode, description, languages[0], country, social.facebook_url, social.twitter_url, social.linkedin_url, profile_url, social.other_urls[0], social.other_urls[1], image_url_valid, image_url_missing_protocol, image_url_invalid_chars, image_url_long, valid_email, domain, valid_url, website_normalized, profile_url_normalized, urls_match, website_valid, website_missing_protocol, website_invalid_chars, website_long, social.facebook_url_valid, social.facebook_url_missing_protocol, social.facebook_url_invalid_chars, social.facebook_url_long]
Index: []

[0 rows x 42 columns]

This has missing email, because it is their home page, where you can find agents and all of their services.

In [82]:
# Regular expression for validating an Email
email_regex = re.compile(r'^[\w\.-]+@[\w\.-]+\.\w+$')

# Function to validate email format
def is_valid_email(email):
    # Convert non-string values to string and handle exceptions
    if isinstance(email, str):
        return bool(email_regex.match(email))
    return False

# Apply validation function
df['valid_email'] = df['email'].apply(is_valid_email)

# Exclude rows with NaN values in 'email' column
df_filtered = df[df['email'].notna()]

# Filter invalid emails, excluding NaN values
invalid_emails = df_filtered[~df_filtered['valid_email']]

print("Invalid email formats:")
print(invalid_emails[['email']])  # Adjust as needed to show relevant columns

Invalid email formats:
  email
6   nan


In [83]:
# Extract domains from email addresses
df['domain'] = df['email'].str.split('@').str[1]

# Count occurrences of each domain
domain_counts = df['domain'].value_counts()

print("Email domain counts:")
print(domain_counts)

Email domain counts:
domain
homerealestate.com    141
HomeRealEstate.com     10
searcyteam.com          3
homeservicesne.com      2
aol.com                 1
gmail.com               1
slamahomes.com          1
HomeServicesne.com      1
Name: count, dtype: int64


Email domains such as 'homerealestate.com' and 'HomeRealEstate.com' are treated as different due to case differences, although they refer to the same domain.

In [84]:
# Convert non-string values to empty strings
df['email'] = df['email'].astype(str)

# Identify emails with common mistakes
# Check for missing '@' symbol and invalid characters
common_mistakes = df[
    ~df['email'].str.contains('@', na=False) |  # No '@' symbol
    df['email'].str.contains(r'[^@\w\.-]', regex=True, na=False)  # Invalid characters
]

print("Emails with common mistakes:")
common_mistakes

Emails with common mistakes:


_id first_name middle_name  last_name  \
6  66cbff36e46b6066699d13bb        Bob         NaN  Tallichet   

             title                                          image_url  \
6  Sales Associate  https://content.mediastg.net/Dynamic/RealEstat...   

  agent_phone_numbers[0] office_phone_numbers[0]                  office_name  \
6           402-430-8255                     NaN  HOME Real Estate At Midtown   

         address  ...                   profile_url_normalized urls_match  \
6  5322 O Street  ...  https://bobtallichet.homerealestate.com       True   

  website_valid website_missing_protocol website_invalid_chars website_long  \
6          True                    False                 False        False   

  social.facebook_url_valid social.facebook_url_missing_protocol  \
6                      True                                False   

  social.facebook_url_invalid_chars social.facebook_url_long  
6                             False                    False  

[1 rows x 42 columns]

## Website

In [86]:
website_qa_results = url_qa(df, 'website')

print("\nWebsite QA Results:")
for key, result_df in website_qa_results.items():
    print(f"\n{key}:")
    # Exclude rows where 'image_url' is NaN
    print(result_df.dropna(subset=['website']))


Website QA Results:

Missing Protocols:
Empty DataFrame
Columns: [website]
Index: []

Invalid Characters:
Empty DataFrame
Columns: [website]
Index: []

Long URLs:
Empty DataFrame
Columns: [website]
Index: []

Invalid URLs:
Empty DataFrame
Columns: [website]
Index: []


## Profile

In [87]:
# Normalize URLs by converting them to lowercase
df['website_normalized'] = df['website'].str.lower()
df['profile_url_normalized'] = df['profile_url'].str.lower()

# Compare the normalized columns
df['urls_match'] = df['website_normalized'] == df['profile_url_normalized']

# Identify rows where URLs do not match
mismatched_urls = df[~df['urls_match']]

# Identify rows where URLs match
matched_urls = df[df['urls_match']]

print("Rows with matching URLs:")
matched_urls

Rows with matching URLs:


_id               first_name middle_name  \
0    66cbff36e46b6066699d13b5  Gregg & Debbie Boosalis         NaN   
1    66cbff36e46b6066699d13b6                      Mai         NaN   
2    66cbff36e46b6066699d13b7                     Jeff         NaN   
3    66cbff36e46b6066699d13b8                      Lea         Ann   
4    66cbff36e46b6066699d13b9                  Jeanine         NaN   
..                        ...                      ...         ...   
156  66cbff36e46b6066699d1451                   Rachel     Roberts   
157  66cbff36e46b6066699d1452                   Marsha         NaN   
158  66cbff36e46b6066699d1453                    Jason         NaN   
159  66cbff36e46b6066699d1454                     Jeff         NaN   
160  66cbff36e46b6066699d1455                   Alicia         NaN   

      last_name            title  \
0           NaN     Team Website   
1            Do          Trainer   
2        Searcy  The Searcy Team   
3    Stephenson         REALTOR®   
4        Curtis  Sales Associate   
..          ...              ...   
156      Spring         REALTOR®   
157        Artz  Sales Associate   
158      Nelson         REALTOR®   
159     Johnson  Sales Associate   
160     Prewett        REALTOR ®   

                                             image_url agent_phone_numbers[0]  \
0    https://content.mediastg.net/Dynamic/RealEstat...           402-432-3666   
1    https://content.mediastg.net/Dynamic/RealEstat...             4023049394   
2    https://content.mediastg.net/Dynamic/RealEstat...           402-430-7978   
3    https://content.mediastg.net/Dynamic/RealEstat...           402-506-3875   
4    https://content.mediastg.net/Dynamic/RealEstat...           402-429-1200   
..                                                 ...                    ...   
156  https://content.mediastg.net/Dynamic/RealEstat...           402-440-8129   
157  https://content.mediastg.net/Dynamic/RealEstat...           402-525-1215   
158  https://content.mediastg.net/Dynamic/RealEstat...           402-219-2505   
159  https://content.mediastg.net/Dynamic/RealEstat...           402-432-0756   
160  https://content.mediastg.net/Dynamic/RealEstat...           402-432-5864   

    office_phone_numbers[0]                  office_name        address  ...  \
0              402-436-4663  HOME Real Estate At Midtown  5322 O Street  ...   
1              402-436-4663  HOME Real Estate At Midtown  5322 O Street  ...   
2                       NaN  HOME Real Estate At Midtown  5322 O Street  ...   
3                       NaN  HOME Real Estate At Midtown  5322 O Street  ...   
4                       NaN  HOME Real Estate At Midtown  5322 O Street  ...   
..                      ...                          ...            ...  ...   
156                     NaN  HOME Real Estate At Midtown  5322 O Street  ...   
157                     NaN  HOME Real Estate At Midtown  5322 O Street  ...   
158                     NaN  HOME Real Estate At Midtown  5322 O Street  ...   
159                     NaN  HOME Real Estate At Midtown  5322 O Street  ...   
160                     NaN  HOME Real Estate At Midtown  5322 O Street  ...   

                                profile_url_normalized urls_match  \
0    https://gregganddebbieboosalis.homerealestate.com       True   
1                     https://maido.homerealestate.com       True   
2                https://jeffsearcy.homerealestate.com       True   
3          https://leaannstephenson.homerealestate.com       True   
4             https://jeaninecurtis.homerealestate.com       True   
..                                                 ...        ...   
156            https://rachelspring.homerealestate.com       True   
157              https://marshaartz.homerealestate.com       True   
158             https://jasonnelson.homerealestate.com       True   
159             https://jeffjohnson.homerealestate.com       True   
160           https://aliciaprewett.homerealestate.com     

In [88]:
print("\nRows with mismatched URLs:")
mismatched_urls


Rows with mismatched URLs:


_id first_name middle_name last_name     title  \
139  66cbff36e46b6066699d1440    Jerilyn         NaN   Hartwig  REALTOR®   

                                             image_url agent_phone_numbers[0]  \
139  https://content.mediastg.net/Dynamic/RealEstat...           402-314-2812   

    office_phone_numbers[0]                  office_name        address  ...  \
139          (402) 436-4663  HOME Real Estate At Midtown  5322 O Street  ...   

                          profile_url_normalized urls_match website_valid  \
139  httppps://jerilynhartwig.homerealestate.com      False          True   

    website_missing_protocol website_invalid_chars website_long  \
139                    False                 False        False   

    social.facebook_url_valid social.facebook_url_missing_protocol  \
139                     False                                 True   

    social.facebook_url_invalid_chars social.facebook_url_long  
139                             False                    False  

[1 rows x 42 columns]

## City

In [89]:
df['city'].value_counts()

city
Lincoln     159
Lincoln       2
Name: count, dtype: int64

Inconsistent Formatting: The city column includes "Lincoln " (with trailing whitespace) and "Lincoln" (without). 

## State

In [90]:
df['state'].value_counts()

state
NE    161
Name: count, dtype: int64

## Zip Code

In [91]:
df['zipcode'].value_counts()

zipcode
68510     158
68506       2
frgthy      1
Name: count, dtype: int64

Invalid Entry: frgthy is an invalid zip code and appears once, indicating data entry errors.
Object Datatype: zip code should be in numerical datatype, but it is object datatype

## Description

In [117]:
df['description'].value_counts()

description
As part of HOME Real Estate, we have a whole team of professionals to help Lincoln home buyers and sellers. That and our HOME-town service are why weve been voted Best of Lincoln every year for more than a decade. Interested in homes for sale or finding out what your home is worth? Register here or contact us today.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               

In [118]:
# Step 1: Find the most common description in the 'title' column
most_common_title = df['description'].mode()[0]

# Step 2: Filter the DataFrame for rows with this description
most_common_title_rows = df[df['description'] == most_common_title]

# Step 3: Print the 'website' column for these rows
print("Websites for rows with the most common title:")
print(most_common_title_rows['website'])


Websites for rows with the most common title:
1                       https://maido.homerealestate.com
4               https://JeanineCurtis.homerealestate.com
12                https://BernieDivis.homerealestate.com
13                   https://DebRaatz.homerealestate.com
15                 https://ClientCare.homerealestate.com
18              https://LeighunBrabec.homerealestate.com
19                https://dmitryrybak.homerealestate.com
20                   https://MNielsen.homerealestate.com
21               https://DeonneHerman.homerealestate.com
23     https://KimMarieSchraderEberspacher.homereales...
24             https://DebbieMessineo.homerealestate.com
29                 https://pamleupold.homerealestate.com
32                https://ValJavorsky.homerealestate.com
33            https://ashleyrockemann.homerealestate.com
39            https://SteveHeironymus.homerealestate.com
41             https://KarenSchaeffer.homerealestate.com
43                https://KatieBugbee.home

## Languages[0]

In [92]:
df['languages[0]'].value_counts()

languages[0]
Russian               20
English               10
Arabic                10
Russian, Ukrainian    10
Greek                  9
Name: count, dtype: int64

Mixed Entries: The languages[0] column contains entries with multiple languages combined (e.g., "Russian, Ukrainian") and single languages (e.g., "Russian").

## Country

In [93]:
df['country'].value_counts()

country
United States    161
Name: count, dtype: int64

## Social.facebook_url

In [94]:
facebook_qa_results = url_qa(df, 'social.facebook_url')

print("\nfacebook_qa_results QA Results:")
for key, result_df in facebook_qa_results.items():
    print(f"\n{key}:")
    # Exclude rows where 'facebook' is NaN
    print(result_df.dropna(subset=['social.facebook_url']))


facebook_qa_results QA Results:

Missing Protocols:
    social.facebook_url
0                   nan
1                   nan
4                   nan
9                   nan
10                  nan
..                  ...
152                 nan
156                 nan
158                 nan
159                 nan
160                 nan

[102 rows x 1 columns]

Invalid Characters:
                                   social.facebook_url
14   https://www.facebook.com/nickhardthomes/?ref=p...
18   https://www.facebook.com//Leighun-Brabec-REALT...
27   https://www.facebook.com/profile.php?id=100039...
40   https://www.facebook.com/profile.php?id=100088...
71   https://www.facebook.com/profile.php?id=100086...
74   https://www.facebook.com/SusanneGuthrie@HOMEwi...
93   https://www.facebook.com/profile.php?id=100095...
130  https://www.facebook.com/profile.php?id=100039...
132  https://www.facebook.com/profile.php?id=100084...
153  https://www.facebook.com/profile.php?id=615567...
155  http

In [96]:
df['social.facebook_url'].value_counts()

social.facebook_url
nan                                                                                                 102
https://www.facebook.com/abbyburmeisterandtiffanyheier                                                4
https://www.facebook.com/profile.php?id=100039183200174                                               3
https://www.facebook.com/DariaandMarsha                                                               3
https://www.facebook.com/Katie-and-Brad-Ulrich-Home-Real-Estate-242594076285712/                      2
https://www.facebook.com/RealtorRoma                                                                  2
https://www.facebook.com/LeaAnnStephensonRealEstate                                                   1
https://www.facebook.com/nickhardthomes/?ref=page_internal                                            1
https://www.facebook.com/SearcyTeamHomeRealEstate                                                     1
https://www.facebook.com/TheSearcyTeamofHOME

## Twitter URL

In [97]:
df['social.twitter_url'].value_counts()

social.twitter_url
https://twitter.com/@Abby_Tiff    2
Name: count, dtype: int64

In [99]:
twitter_url_qa_results = url_qa(df, 'social.twitter_url')

print("\ntwitter_url_qa_results QA Results:")
for key, result_df in twitter_url_qa_results.items():
    print(f"\n{key}:")
    # Exclude rows where 'facebook' is NaN
    print(result_df.dropna(subset=['social.twitter_url']))


twitter_url_qa_results QA Results:

Missing Protocols:
    social.twitter_url
0                  nan
1                  nan
2                  nan
3                  nan
4                  nan
..                 ...
156                nan
157                nan
158                nan
159                nan
160                nan

[159 rows x 1 columns]

Invalid Characters:
                 social.twitter_url
77   https://twitter.com/@Abby_Tiff
108  https://twitter.com/@Abby_Tiff

Long URLs:
Empty DataFrame
Columns: [social.twitter_url]
Index: []

Invalid URLs:
    social.twitter_url
0                  nan
1                  nan
2                  nan
3                  nan
4                  nan
..                 ...
156                nan
157                nan
158                nan
159                nan
160                nan

[159 rows x 1 columns]


## Linkedin URL

In [103]:
linkedin_url_counts= df['social.linkedin_url'].value_counts()

In [105]:
# Step 2: Filter for URLs that appear more than once
duplicate_linkedin_urls = linkedin_url_counts[linkedin_url_counts > 1].index

# Step 3: Filter the original DataFrame to get rows with these duplicate URLs
duplicate_linkedin_rows = df[df['social.linkedin_url'].isin(duplicate_linkedin_urls)]

# Print the rows with duplicate 'social.linkedin_url'
print("Rows with duplicate social.linkedin_url:")
print(duplicate_linkedin_rows[['website']])

Rows with duplicate social.linkedin_url:
                                    website
61  https://KatieAndBrad.homerealestate.com
68   https://katieulrich.homerealestate.com


They have a team website and individual website. That's why duplicate linkedin urls

## Other URLs[0]

In [108]:
df['social.other_urls[0]'].value_counts()

social.other_urls[0]
https://www.instagram.com/andie.jessi_realestate/            2
https://www.instagram.com/abby_tiffany_home_real_estate/     2
http://blog.homerealestate.com/                              1
https://www.instagram.com/ellenhighrealtor/?hl=en            1
https://www.instagram.com/dariaandmarsha_home/               1
https://www.instagram.com/heatherfergusonrealtorwithhome/    1
https://www.instagram.com/searcyteam                         1
https://twitter.com/@Abby_Tiff                               1
https://www.instagram.com/nel2sel/                           1
Name: count, dtype: int64

Twitter url is in this column

In [115]:
# Filter the DataFrame for rows where 'social.other_urls' matches the specified URL
matching_rows = df[df['social.other_urls[0]'] == 'https://twitter.com/@Abby_Tiff']

# Print the matching rows
print("Rows with social.other_urls as 'https://twitter.com/@Abby_Tiff':")
matching_rows

Rows with social.other_urls as 'https://twitter.com/@Abby_Tiff':


_id first_name middle_name   last_name  \
65  66cbff36e46b6066699d13f6       Abby         NaN  Burmeister   

              title                                          image_url  \
65  Sales Associate  https://content.mediastg.net/Dynamic/RealEstat...   

   agent_phone_numbers[0] office_phone_numbers[0]  \
65           402-480-1727                     NaN   

                    office_name        address  ... website_invalid_chars  \
65  HOME Real Estate At Midtown  5322 O Street  ...                 False   

   website_long social.facebook_url_valid  \
65        False                      True   

   social.facebook_url_missing_protocol social.facebook_url_invalid_chars  \
65                                False                             False   

   social.facebook_url_long social.twitter_url_valid  \
65                    False                    False   

   social.twitter_url_missing_protocol social.twitter_url_invalid_chars  \
65                                True                            False   

   social.twitter_url_long  
65                   False  

[1 rows x 46 columns]

## Social.other_urls[1]

In [116]:
df['social.other_urls[1]'].value_counts()

social.other_urls[1]
https://www.instagram.com/annie_steph07/                    1
https://www.pinterest.com/tastytaters/                      1
https://www.instagram.com/abby_tiffany_home_real_estate/    1
Name: count, dtype: int64